gc.c

```diff
@@ -8209,6 +8022,172 @@ gc_marks_step(rb_objspace_t *objspace, size_t slots)
 }
 #endif
 
+static bool
+gc_compact_heap_cursors_met_p(rb_heap_t *heap)
+{
+    GC_ASSERT(heap->sweeping_page && heap->compact_cursor);
+    return heap->sweeping_page == heap->compact_cursor;
+}
+
+static void
+gc_compact_sweep_unswept_page(rb_objspace_t *objspace, rb_heap_t *heap, struct heap_page *page)
+{
+    if (page->flags.before_sweep) {
```

Comment:
Why do we need to check `before_sweep`? Shouldn't sweeping pages be pulled from the `heap->sweeping_page` cursor which guarantee this?

Reply:
yes, but in the current implementation, we ensure the page has been swept before attempting to move an object into it. `try_move` will be called with multiple `src` objects, but the same `sweeping_page` and we want to ensure that we only try and sweep the page once.